ADR/0027-availability-probe-framework.md

```diff
@@ -0,0 +1,125 @@
+# 27. Availability Probe Framework
+
+* Date 2023-07-06
+
+## Status
+
+Accepted
+
+## Context
+
+As an AppStudio developer building functionality for the platform, I want to be able to
+easily visualize and comprehend the stability and availability of deployed systems in
+order to inform and influence future work towards improving the overall system
+reliability.
+  
+Such indication should tell us the overall up time of AppStudio with respect to services
+under the control of AppStudio developers.
+
+AppStudio is defined to be available at a given moment if all of its components are
+reporting to be available at that given moment.
+
+A component is defined to be available at a given moment if all of its availability
+probes are reporting to be available at that given moment.
+
+A convention is required for providing the availability of a probe.
+
+Once this is in place, those indicators can be aggregated in order to report the overall
+availability of AppStudio.
+
+## Decision
+
+Probes' availability will be provided as a Prometheus metric. The metric will contain
+metric labels to allow differentiating between the different probes. When exported out
```

Comment:
If there are multiple probes per component, you need to be explicit about these probes -- whether it is differentiating between any probe or between any given component's probe(s).

Reply:
> Is it valid for a single component to have multiple probes?
It's valid. Team should create as many probes as they see fit.

Reply:
I understand that you want to keep the paragraph brief, but I think that we need to ensure that the language is precise when it comes to different probes. Probes are not always on a per component basis, so if you are referencing the probes for a specific component, it needs to be plural. Also when referencing the aggregated availability, it isn't if a specific component reports being unavailable, but a specific probe.

Reply:
I can't say I understand. Should we say here that components might have one or more probes?

Reply:
Yes, that is what you are telling me. I am just trying to say that we need to ensure that we are clearly communicating this relationship and that we need to ensure that we are properly considering all cases where assumptions might fall apart or be unnecessary.

For example, later when you talk about components, you cannot talk about a single probe per component. Also when discussing the probe names, what might it look like for a service which has multiple components and some components have multiple probes? Trying to differentiate between a unique probe name and a unique component name doesn't fit my mental model. Instead, would it just make sense to describe only as a probe name and indicate that some use cases might see a single probe being able to represent the status for a single component? Would there be any ability to see an aggregation of probe metrics on a component level instead of on an AppStudio level? If so, is there any abstraction that can be provided? Should the probe naming convention be scoped to have the name include the component and a probe identifier?

Reply:
We're saying this:
```
The nature and size of each service and component will dictate the number of probes it
should have. E.g. some services may have multiple components, while some others may have
just one. Some components may require multiple probes while others may require just one.
```

Can you point me to places where we have probe mentioned in a way that may suggest different probe-component cardinality?

>  what might it look like for a service which has multiple components and some components have multiple probes?

It will have namespaces defined per component, under each of them there will be cronjobs defined per probe.

> Would there be any ability to see an aggregation of probe metrics on a component level instead of on an AppStudio level?

That's the idea, and it's also being mentioned: `To allow aggregating the Prometheus availability metric per component...`.

> If so, is there any abstraction that can be provided?

Creating the cronjobs for a single component on a single namespace is the means to enable that.

> Should the probe naming convention be scoped to have the name include the component and a probe identifier?

I didn't want to make the teams have to coordinate probe names between the entire appstudio. For that reason, the requirement is to have the identifier of the probe within the cronjob name, and have that multiplexed by the namespace identifying the component.

Reply:
@arewm let me know your thoughts.